qiskit/circuit/library/n_local/qaoa_ansatz.py

```diff
@@ -0,0 +1,258 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2018, 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A generalized QAOA quantum circuit with a support of custom initial states and mixers."""
+# pylint: disable=R0401
+from typing import Optional, cast, Set
+
+import numpy as np
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.parametervector import ParameterVector
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from ..blueprintcircuit import BlueprintCircuit
+
+
+class QAOAAnsatz(BlueprintCircuit):
+    """A generalized QAOA quantum circuit with a support of custom initial states and mixers.
+
+    References:
+
+        [1]: Farhi et al., A Quantum Approximate Optimization Algorithm.
+            `arXiv:1411.4028 <https://arxiv.org/pdf/1411.4028>`_
+    """
+    def __init__(self,
+                 cost_operator,
+                 reps: int,
```

Comment:
I cannot see the comments I thought I made before. Perhaps I lost the comment by not adding it. Anyway sorry if this is repeated.

Lets do reps comment first as its simpler. In the other blueprint circuits like NLocal, TwoLocal there reps has a default - can we not default it to 1 here so these ansatz Blueprint circuits are more consistent.

Then for cost_operator, can we default it to None so it does not have to be set up front but can be set later. This makes it more akin to the other blueprint circuits that are templates that need num_qubits and where this defaults to None upfront and can be set later. By having this as None it seems to feel more like the template notion. This allows it to be more flexible so a user can create a QAOAAnsatz not knowing the operator and pass it to an algorithm that will later set the operator in the same manner as we do today with NLocal ccts. Now QAOA today does not work this was and has it as an internal ansatz and can set it up front. Which is fine. But it makes the QAOAAnsatz more flexible for future usage and makes it more consistent in behavoir in these regards with the others there.

Reply:
@woodsp-ibm, yes `reps=1` makes sense. Optional cost operator is fine also. I will do this.